File: source/wave.c
#include <stddef.h>
#include <malloc.h>
#include <string.h>
#include <stdlib.h> 
#include <stdbool.h>
#include <math.h>
#include "constants.h"
#include "geometry.h"
#include "regular_wave.h"
#include "wave.h"
#include "errors.h"

struct Wave
{
  // Input variables
  // ---------------
  double significant_wave_height;      //!< Input variable. Significant wave height in meter.
  double heading;                      //!< Input variable. Wave heading in radians.
  long random_number_seed;             //!< Input variable. Random number generator seed. 
  int count_wave_spectral_directions;  //!< Input variable. Number of direction bands in the wave spectrum. 
  int count_wave_spectral_frequencies; //!< Input variable. Number of frequency bands in the wave spectrum.

  // Output variables
  // ----------------
  struct Regular_wave** spectrum;   //!< Output variable. Table of regular waves in the irregular sea.
                                           //!< Size is count_wave_spectral_directions * count_wave_spectral_frequencies.
  double min_spectral_frequency;    //!< Output variable. Lower limit (0.1%) of spectral energy threshold.
  double max_spectral_frequency;    //!< Output variable. Upper limit (99.9%) of spectral energy threshold.
  double peak_spectral_frequency;   //!< Output variable. Spectral peak frequency in Hz.
  double min_spectral_wave_heading; //!< Output variable. Minimum angle, in radians, in spectrum for wave heading.
  double max_spectral_wave_heading; //!< Output variable. Maximum angle, in radians, in spectrum for wave heading.
  char* error_msg;                  //!< Output variable. Error message, if any.
};

const char* wave_get_error_msg(const struct Wave* wave)
{
  if(wave)
  {
    return wave->error_msg;
  }
  return NULL;
}


struct Wave* wave_new(const double sig_wave_ht,
                      const double wave_heading, 
                      const long rand_seed,
                      const int count_wave_spectral_directions,
                      const int count_wave_spectral_frequencies)
{
  struct Wave* wave = NULL;

  if(sig_wave_ht > 0.0 && 
     count_wave_spectral_directions > 1 && 
     count_wave_spectral_frequencies > 1)
  {
    // Initialise the pointers...
    if(wave = (struct Wave*)malloc(sizeof(struct Wave)))
    {
      if(wave->spectrum = (struct Regular_wave**)malloc(sizeof(struct Regular_wave*) * count_wave_spectral_directions * count_wave_spectral_frequencies)) 
      {
        wave->error_msg = NULL;
        // ... and then the other member variables.
        wave->random_number_seed = rand_seed;
        srand(wave->random_number_seed);
        wave->heading = wave_heading;
        wave->count_wave_spectral_directions = count_wave_spectral_directions;
        wave->count_wave_spectral_frequencies = count_wave_spectral_frequencies;
        wave->min_spectral_wave_heading = wave->heading - PI/2.0;
        wave->max_spectral_wave_heading = wave->heading + PI/2.0;
      
        // wave directions should be in the range (0, 2PI)
        wave->min_spectral_wave_heading = fmod(wave->min_spectral_wave_heading,  2.0*PI); 
        wave->max_spectral_wave_heading = fmod(wave->max_spectral_wave_heading , 2.0*PI);
        
        // Bretschneider spectrum
        // Ref: Proceedings of the 23rd ITTC - Vol II, Table A.2, A.3.
        // S(f) = (A/f^5) exp(-B/f^4)
        // A = alpha g^2 (2 PI)^-4
        // B = beta (2PI U/g)^-4
        // alpha = 0.0081
        // beta = 0.74
        // f_p = 0.946 B^(1/4)
        // U = wind speed in m/s
        double alpha = 0.0081;
        double beta = 0.74;
        double A = alpha * G*G * pow(2.0*PI, -4.0);
        double H_s = sig_wave_ht;
        double B = 4.0*alpha*G*G / (pow(2.0*PI, 4.0)* H_s*H_s);
        double f_p = 0.946 * pow(B, 0.25);

        wave->significant_wave_height = H_s;
        wave->peak_spectral_frequency = f_p;
        wave->min_spectral_frequency = 0.652 * f_p;
        wave->max_spectral_frequency = 5.946 * f_p;

        // Create regular waves
        bool has_NULL_in_spectrum = false; // A flag that will be set to true if any of the created component wave turns out to be NULL.
        // For each heading angle
        double wave_heading_step_size = PI/(count_wave_spectral_frequencies - 1);
        double mu = -PI/2.0;
        for(int i = 0; i < count_wave_spectral_directions; ++i)
        {
          mu += wave_heading_step_size;
          double frequency_step_size = (wave->max_spectral_frequency - 
                                        wave->min_spectral_frequency) /
                                        (count_wave_spectral_frequencies - 1);
          for(int j = 0; j < count_wave_spectral_frequencies; ++j)
          {
            double f = wave->min_spectral_frequency + j * frequency_step_size;
            double S = (A/pow(f,5.0)) * exp(-B/pow(f,4.0)) * frequency_step_size;
          
            // Direction function G = (2/PI) * cos(mu)*cos(mu) * delta_mu
            // delta_mu = wave_heading_step_size
            double G_spectrum = (2.0/PI) * cos(mu)*cos(mu) * wave_heading_step_size;
          
            // Create a wave
            double amplitude = sqrt(2.0 * S * G_spectrum); 
            double phase = rand(); 
            double wave_heading = mu + wave->heading;
            // wave directions should be in the range (0, 2PI)
            wave_heading = fmod(wave_heading, 2.0 * PI);
            struct Regular_wave* regular_wave = regular_wave_new(amplitude, f, phase, wave_heading);
            if(regular_wave)
            {
              wave->spectrum[i*count_wave_spectral_frequencies + j] = regular_wave;
            }
            else 
            {
              // Error encountered when creating a regular waves for the wave spectrum.
              has_NULL_in_spectrum = true;
              break;
            }
          }
          if(has_NULL_in_spectrum)
          {
            break;
          }
        }
        // Check if spectrum is valid
        if(has_NULL_in_spectrum)
        {
          // One or more of the regular wave is NULL. 
          // This is an invalid spectrum.
          wave_delete(wave);
          wave = NULL;
        }
      }
      else
      {
        free(wave);
        wave = NULL;
      }
    }
  }
  
  return wave;
}

void wave_delete(struct Wave* wave)
{
  if(wave)
  {
    for(int i = 0; i < (wave->count_wave_spectral_directions * wave->count_wave_spectral_frequencies); ++i)
    {
      if(wave->spectrum[i])
      {
        regular_wave_delete(wave->spectrum[i]);
      }
    }
    
    free(wave->spectrum);
    free(wave->error_msg);
    free(wave);
    wave = NULL;
  }
}

double wave_get_elevation(const struct Wave* wave, 
                          const union Coordinates_3D location,
                          double time)
{
  clear_error_msg(wave->error_msg);
  double elevation = 0.0;

  if(wave)
  {
    // check if time is negative
    if(time >= 0.0)
    {
      for(int i = 0; i < wave->count_wave_spectral_directions; ++i)
      {
        for(int j = 0; j < wave->count_wave_spectral_frequencies; ++j)
        {
          const struct Regular_wave* regular_wave = wave_get_regular_wave_at(wave, i, j);
          if(!regular_wave)
          {
            // Something really wrong happened.
            // Error should already be set by wave_get_regular_wave_at().
            return 0.0;
          }
          double regular_wave_elevation = regular_wave_get_elevation(regular_wave, location, time);
          const char* error_msg = regular_wave_get_error_msg(regular_wave);
          if(error_msg)
          {
            // Something went wrong when getting wave elevation for the regular_wave.
            set_error_msg(wave->error_msg, error_msg);
            return 0.0;
          }
          else
          {
            elevation += regular_wave_elevation;
          }
        }
      }
    }
    else
    {
      set_error_msg(wave->error_msg, error_negative_time);
      return 0.0;
    }
  }
  else
  {
    set_error_msg(wave->error_msg, error_null_pointer);
    return 0.0;
  }
  return elevation;
}

int wave_get_count_wave_spectral_directions(const struct Wave* wave)
{
  clear_error_msg(wave->error_msg);
  if(wave)
  {
    return wave->count_wave_spectral_directions;
  }
  else
  {
    set_error_msg(wave->error_msg, error_null_pointer);
    return 0.0;
  }
}

int wave_get_count_wave_spectral_frequencies(const struct Wave* wave)
{
  clear_error_msg(wave->error_msg);
  if(wave)
  {
    return wave->count_wave_spectral_frequencies;
  }
  else
  {
    set_error_msg(wave->error_msg, error_null_pointer);
    return 0.0;
  }
}

const struct Regular_wave* wave_get_regular_wave_at(const struct Wave* wave, int d, int f)
{
  clear_error_msg(wave->error_msg);
  if(wave)
  {
    if(d >= 0 && d < wave->count_wave_spectral_directions &&
      f >= 0 && f < wave->count_wave_spectral_frequencies)
    {
      return wave->spectrum[d*wave->count_wave_spectral_frequencies + f];
    }
    else
    {
      // Incorrect index
      set_error_msg(wave->error_msg, error_invalid_index);
      return NULL;
    }
  }
  else
  {
    set_error_msg(wave->error_msg, error_null_pointer);
    return NULL;
  }
}

double wave_get_min_spectral_frequency(const struct Wave* wave)
{
  clear_error_msg(wave->error_msg);
  if(wave)
  {
    return wave->min_spectral_frequency;
  }
  else
  {
    set_error_msg(wave->error_msg, error_null_pointer);
    return 0.0;
  }
}

double wave_get_max_spectral_frequency(const struct Wave* wave)
{
  clear_error_msg(wave->error_msg);
  if(wave)
  {
    return wave->max_spectral_frequency;
  }
  else
  {
    set_error_msg(wave->error_msg, error_null_pointer);
    return 0.0;
  }
}

double wave_get_significant_height(const struct Wave* wave)
{
  clear_error_msg(wave->error_msg);
  if(wave)
  {
    return wave->significant_wave_height;
  }
  else
  {
    set_error_msg(wave->error_msg, error_null_pointer);
    return 0.0;
  }
}

double wave_get_predominant_heading(const struct Wave* wave)
{
  clear_error_msg(wave->error_msg);
  if(wave)
  {
    return wave->heading;
  }
  else
  {
    set_error_msg(wave->error_msg, error_null_pointer);
    return 0.0;
  }
}

File: ASVLite_Rust/src/wave.rs
//! Implementation of an irregular wave.

use libc;
use super::error::ValueError;
use super::geometry::Dimension;
use super::regular_wave::RegularWave;
use super::constants::{PI, G};

extern {
    fn srand(seed: libc::c_uint);
    fn rand() -> libc::c_int;
}

/// Structure to define an irregular wave. The members of the struct is initialized 
/// by the associated function new(). The member variables are all public and can be directly
/// accessed for reading. However, the member variables should not be directly accessed for writing. 
/// The struct does not provide any member function for editing the values of a member variable after 
/// the value is set. So to edit a member variable value, drop the current instance of the struct and 
/// create a new one with the desired value. Recommend assigning the return of new() to an immutable variable.
#[derive(Debug, Clone, PartialEq)]
pub struct Wave {
    /// Significant wave height in meter.
    pub significant_wave_height: f64,
    /// Wave heading in radians.
    pub heading: f64,
    /// Random number generator seed.
    pub random_number_seed: u32,
    /// Lower limit (0.1%) of the spectral energy threshold.
    pub min_spectral_frequency: f64,
    /// Upper limit (99.9%) of the spectral energy threshold.
    pub max_spectral_frequency: f64,
    /// Spectral peak frequency in Hz.
    pub peak_spectral_frequency: f64,
    /// Minimum angle, in radians, for wave heading in the spectrum.
    pub min_spectral_wave_heading: f64,
    /// Maximum angle, in radians, for wave heading in the spectrum. 
    pub max_spectral_wave_heading: f64,
    /// Wave spectrum
    pub spectrum: Vec<Vec<RegularWave>>,
}

impl Wave {
    /// Initialize a new irregular wave.
    pub fn new(significant_wave_height: f64, 
                heading: f64, 
                random_number_seed: u32, 
                count_spectral_directions: usize, 
                count_spectral_frequencies: usize) -> Result<Wave,ValueError> {
        match significant_wave_height <= 0.0 {
            true => Err(ValueError::new("Significant wave height should not be <= 0.0.")),
            false => {
                let mut min_spectral_wave_heading = heading - PI/2.0;
                let mut max_spectral_wave_heading = heading + PI/2.0;
                // wave directions should be in the range (0, 2PI)
                if min_spectral_wave_heading < 0.0 {
                    min_spectral_wave_heading += 2.0*PI; 
                }
                if max_spectral_wave_heading >= 2.0*PI {
                    max_spectral_wave_heading -= 2.0*PI;
                }

                // Bretschneider spectrum
                // Ref: Proceedings of the 23rd ITTC - Vol II, Table A.2, A.3.
                // S(f) = (A/f^5) exp(-B/f^4)
                // A = alpha g^2 (2 PI)^-4
                // B = 4 alpha g^2 / [(2 PI)^4 H_s^2]
                // alpha = 0.0081
                // beta = 0.74
                // f_p = 0.946 B^(1/4)
                // U = wind speed in m/s
                let alpha = 0.0081;
                // let beta = 0.74;
                let a = alpha * G*G * (2.0 * PI).powf(-4.0);
                let h_s = significant_wave_height;
                let b = 4.0*alpha*G*G / ((2.0*PI).powf(4.0) * h_s*h_s);
                let f_p = 0.946 * b.powf(0.25);
                let min_spectral_frequency = 0.652 * f_p;
                let max_spectral_frequency = 5.946 * f_p;

                // Create the regular waves
                let wave_0 = RegularWave::new(1.0, 1.0, 0.0, 0.0).unwrap(); // A temp wave to initialise the spectrum.
                let mut spectrum = vec![vec![wave_0; count_spectral_frequencies]; count_spectral_directions];
                let wave_heading_step_size = PI/((count_spectral_directions as f64) - 1.0);
                let frequency_step_size = (max_spectral_frequency - min_spectral_frequency) /
                                          ((count_spectral_frequencies as f64) - 1.0);
                let mut mu = -PI/2.0;
                for i in 0 .. count_spectral_directions {
                    mu += wave_heading_step_size;
                    for j in 0 .. count_spectral_frequencies {
                        let f = min_spectral_frequency + (j as f64) * frequency_step_size;
                        let s = (a/f.powf(5.0)) * (-b/f.powf(4.0)).exp() * frequency_step_size;

                        // Direction function G = (2/PI) * cos(mu)*cos(mu) * delta_mu
                        // delta_mu = wave_heading_step_size
                        let g_spectrum = (2.0/PI) * mu.cos()*mu.cos() * wave_heading_step_size;

                        // Create a wave
                        let amplitude = (2.0 * s * g_spectrum).sqrt(); 
                        unsafe{ srand(random_number_seed) };
                        let phase = unsafe{ rand() } as f64;
                        let mut wave_heading = mu + heading;
                        // wave directions should be in the range (0, 2PI)
                        while wave_heading < 0.0 {
                            wave_heading += 2.0 * PI; 
                        }
                        while wave_heading >= 2.0*PI {
                            wave_heading -= 2.0*PI;
                        }
                        spectrum[i][j] = RegularWave::new(amplitude, f, phase, wave_heading).unwrap();
                    }
                }

                Ok(
                    Wave{
                    significant_wave_height: h_s,
                    heading,
                    random_number_seed,
                    spectrum,
                    min_spectral_frequency,
                    max_spectral_frequency,
                    peak_spectral_frequency: f_p,
                    min_spectral_wave_heading,
                    max_spectral_wave_heading,
                    }
                )
            },
        }
    }

    /// Get wave elevation at a given location at a given time. 
    pub fn get_elevation(&self, location: &Dimension, time: f64) -> Result<f64,ValueError> {
        match time < 0.0 {
            true => Err(ValueError::new("Time should not be < 0.0.")),
            false => {
                let mut elevation = 0.0;
                let count_spectral_directions = self.spectrum.len();
                for i in 0 .. count_spectral_directions {
                    let count_spectral_frequencies = self.spectrum[i].len();
                    for j in 0 .. count_spectral_frequencies {
                        elevation += self.spectrum[i][j].get_elevation(location, time).unwrap();
                    }
                }
                Ok(elevation)
            },
        }
    }
}